### PR TITLE
build: enable more strictness flags in non-library code

### PR DESCRIPTION
**Pull request by crisbeto**
Updates the compiler options for the schematics and other Node-based code, and resolves any compilation errors that showed up as a result. Also fixes a bunch of references to non-existent types from `parse5`.

### CODE REVIEW

File: src/cdk/schematics/ng-update/migrations/attribute-selectors.ts [hunk 1]
```diff
@@ -23,23 +23,23 @@ export class AttributeSelectorsMigration extends Migration<UpgradeData> {
   data = getVersionUpgradeData(this, 'attributeSelectors');
 
   // Only enable the migration rule if there is upgrade data.
-  enabled = this.data.length !== 0;
+  enabled: boolean = this.data.length !== 0;
```
Comment (devversion) [inline src/cdk/schematics/ng-update/migrations/attribute-selectors.ts:26]: I wonder why this is needed. Seems like something TS could infer quite easily?

File: src/cdk/schematics/utils/build-component.ts [hunk 2]
```diff
@@ -175,8 +175,13 @@ export function buildComponent(options: ComponentOptions,
     // Add the default component option values to the options if an option is not explicitly
     // specified but a default component option is available.
     Object.keys(options)
-      .filter(optionName => options[optionName] == null && defaultComponentOptions[optionName])
-      .forEach(optionName => options[optionName] = defaultComponentOptions[optionName]);
+      .filter((optionName: keyof ComponentOptions) => {
+        return options[optionName] == null &&
+               (defaultComponentOptions as ComponentOptions)[optionName];
```
Comment (devversion) [inline src/cdk/schematics/utils/build-component.ts:180]: Have you considered updating `getDefaultComponentOptions` to return a `Partial<ComponentOptions>` or just `ComponentOptions`?

**Review by devversion — APPROVED** [review]
Nice! Ideally in the long-term our nested tsconfig's would all inherit the strictness properties from our project `bazel-tsconfig-build` file, but that is more complicated than it sounds.

File: src/cdk/schematics/ng-update/migrations/attribute-selectors.ts (hunk 1 above)
Comment (crisbeto) [reply src/cdk/schematics/ng-update/migrations/attribute-selectors.ts:26]: It's a weird one, but for some reason TS was flagging the entire class if the type isn't specified explicitly. It was something about not being able to infer the type from the constructor.

**Pull request merged**